app/api/services/users/models.py

```diff
@@ -0,0 +1,24 @@
+import dataclasses
+from datetime import date
+
+from api.db.models import user_models
+from api.services.core import patch_params
+
+
+@dataclasses.dataclass
+class RoleParams:
+    type: user_models.RoleType
+
+
+@dataclasses.dataclass
+class CreateUserParams:
+    first_name: str | None = None
+    middle_name: str | None = None
+    last_name: str | None = None
+    phone_number: str | None = None
+    date_of_birth: date | None = None
+    is_active: bool | None = None
+    roles: list[RoleParams] | None = None
```

Comment:
Hmm, this does make working with the object have one annoyance. Let's assume `first_name` is always required based on the schema. We as developers know this because the validation happened, but in-code we'd have to check that it's not None to make mypy happy. `cast` helps, but it can't trace that through the code very well so anywhere you use it and it being null is a problem, mypy will require you to cast again.

This might be wishful thinking, but it'd be great if you could define a dataclass as:
```
@dataclass
class MyDataclass:
     param1: str
     param2: str | None
     param3: str
```
and then generate another class with identical params, but all of them optional. It might be possible to do that via https://docs.python.org/3/library/dataclasses.html#dataclasses.make_dataclass by iterating over the original classes fields, but is that making things too abstract?

Reply:
if you look in `create_user.py` below, we assert that certain fields aren't `None`, but don't do those assertions in `patch_user.py` since they can be `None` in the patch context. what do you think about that approach? to me it feels like the simplest one to understand and still has benefit of static analysis, since the asserts effectively rule out the None values in the context of create_user. I considered writing out separate classes but that will have a lot of duplicate code, and generating the separate class dynamically doesn't give us the benefit static analysis.

Reply:
It would only help in that particular function. Let's assume that right after you do all of those asserts you do:

```py
@dataclass
class MyModel:
     field: Optional[str] = None

assert my_model.field is not None
persist_my_model(my_model)

def persist_my_model(my_model: MyModel) -> None:
      my_field: str = my_model.field
```
mypy will actually complain because `field` is nullable despite you doing an assert because you could call this persist function from other places. You'd have to do these asserts in every separate part of the code, so they're actually doing nothing.

Reply:
I wonder if there is a way to solve this with generics.


```py
K = TypeVar("K") # This can be either `Optional` or whatever the opposite of that is (Any?)

class MyModel[Generic[K]):
     param1: K[str] # So this is either Optional[str], or DummyWrapper[str] (which is the same as str)
     param2: Optional[str] # this is always optional even in post, so no need to do anything
     param3: Optional[str]
```

We might need to not use dataclasses in this case and instead use pydantic as it handles defaults cleaner.

Reply:
yeah the generic idea is interesting, i had considered using python metaprogramming functionality as well, but worried that it's too complicated

Reply:
If it's reasonable to make the type hint as simple as needing to mark a class as patch or not, then I don't think it'd be TOO complex to understand, but really depends on how much needs to be done for mypy to be happy.

I'm really surprised that there isn't a happy solution online, most of the time people just say to work with dictionaries and I really don't want to see those systems. Best solution I've seen is just defining the model twice which works, is clear, just mildly error-prone/tedious for large objects.

Reply:
yeah i'm surprised too.

i also wonder if we're overcomplicating things. is it really that bad to just have two separate classes e.g.

```
class CreateUserParams:
    first_name: str
    last_name: str

class PatchUserParams:
    first_name: Optional[str]
    last_name: Optional[str]
```

or just define the params directly in the service methods

```
def create_user(
    first_name: str,
    last_name: str
):
    ...

def patch_user(
    first_name: Optional[str],
    last_name: Optional[str]
):
    ...
```

it's very simple and straightforward and to the point.

i often follow the rule of thumb of "three", where i only try to generalize and abstract once we duplicate the same thing 3 times. but create and patch is only two.

thoughts?

Reply:
I think two separate classes is fine. Whatever time we might spend on this to solve with a clever solution at this point probably offsets the small gains from anyone using it later.

The function approach is fine up to a point, if we're talking 3-4 params sure, but if we have a 20+ param then I'd probably make an object to hold all the params for the function anyways (or you can use kwargs which is literally back to being the dict approach).

I think at the end of the day, what I'd do if I were building an API with full control is just never let any of these models get too big. Probably avoid anything like the Mass PATCH application endpoint and instead make those a bit more targeted if possible (per-page on the front-end?)

Reply:
ok pushed up a change with two separate params classes, and also moved them to the create_user and patch_user files respectively so the params live with the service methods.

note that `patch_user.py` imports `create_user.RoleParams`, does that feel ok or weird to you?

Reply:
@chouinar doing some more reading, i think i found something promising.

what do you think about [using a TypedDict](https://docs.python.org/3/library/typing.html#typing.TypedDict) instead of a dataclass?

we could do

```
user_fields = {
    "first_name": str,
    "middle_name": str,
    "last_name": str,
    "phone_number": str,
    "date_of_birth": date,
    "is_active": bool,
    "roles": list[RoleParams],
}

CreateUserParams = TypedDict('CreateUserParams', user_fields)
PatchUserParams = TypedDict('PatchUserParams', user_fields, total=False)
```

the `total=False` marks all the fields as `NotRequired`

Reply:
Let me test this a little, seems promising.

Reply:
A few notes from tinkering with it locally.

My IDE is able to auto-complete when I type `user["` it'll give me a list of options. It also shows type hints properly for `z = user["x"]`.

A bit of tinkering with mypy, it behaves as I would expect. Extra fields in the dictionary give an error. Typing still works (total=False allows it to not appear, but if it is there it has to follow typing).

I'm not finding any issues, I'm onboard with this idea.

Reply:
i tinkered with it a bit too and ran into one issue, i'm wondering if you ran into it too.

when i use the functional syntax for creating the TypedDict

```
user_fields = {
    "first_name": str,
}

CreateUserParams = TypedDict('CreateUserParams', user_fields)

def create_user(params: CreateUserParams):
    params["first_name"] # this raises a type error
```

but when I use the class syntax like below it's ok

```
CreateUserParams(TypedDict):
    first_name: str

def create_user(params: CreateUserParams):
    params["first_name"] # this doesn't raise a mypy error anymore
```

i was hoping the functional syntax could work because then we could reuse the same dictionary to define `CreateUserParams` and `PatchUserParams` except `PatchUserParams` would have `total=False`

did you run into that at all?

otherwise we could just define two types separately with a bit of duplication

Reply:
Doing it the functional way, I'm not getting any different behavior. What type error are you getting?

Reply:
i'm getting

> TypedDict "TypedDict" has no key "first_name"

<img width="888" alt="image" src="https://user-images.githubusercontent.com/447859/210662249-6511d959-4ca8-4e0e-88ba-ded828ea9c1f.png">

Reply:
It looks like it needs to use a literal dict, I can't assign a dict to a variable and then pass in the variable.

e.g. this works
```
Point2D = TypedDict('Point2D', {'x': int, 'y': int, 'label': str})
```

but this doesn't

```
fields = {'x': int, 'y': int, 'label': str}
Point2D = TypedDict('Point2D', fields)
```

Reply:
Ok tried out TypedDict in a separate branch so we can compare the two: https://github.com/navapbc/template-application-flask/pull/62

Take a look and lemme know what you think.

Reply:
Ok I updated the title and description of this PR now that we're using TypedDict instead of objects